policy: cache aggregated list of selectors in rule

jrajahalme opened the pull request:
Add a new, unexported field to the `IngressRule` and `EgressRule` types in the
API called `aggregatedSelectors`.

For `IngressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `FromEntities`
* `FromCIDR`
* `FromCIDRSet`

For `EgressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `ToEntities`
* `ToCIDR`
* `ToCIDRSet`
* `ToFQDNs`

Previously, upon each policy calculation, for each rule, this aggregate list of
rule fields as `EndpointSelector` was created in an ad-hoc manner, and was not
performant, as it allocated many different slices of `EndpointSelector` in calls
to `GetSourceEndpointSelectorsWithRequirements` and
`GetDestinationEndpointSelectorsWithRequirements`. This new field caches the
aggregated list only once, upon rule sanitization, or in calls to the two
aforementioned functions if the field has not been initialized yet.

The performance gains by making this change are significant. With 1000
egress rules containing a ToCIDR with 2 CIDRs, and 3000 identities
from before making this change:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     100	  12630576 ns/op	 4127697 B/op	  106148 allocs/op
```

And after:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     500	   6208546 ns/op	 2265445 B/op	   61134 allocs/op
```

This is a 50 % performance improvement in policy calculation for this simple
policy benchmark.

Note: This is a revival of PR #7250 (commit 21fe2cc18c99 of the same name) that got reverted in f64cf1c420f2 ("policy: Remove more dead code.") due to not being used any more.

Signed-off-by: Ian Vernon <ian@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8282)
<!-- Reviewable:end -->

ianvernon commented on pkg/policy/api/egress.go:
```diff
@@ -148,12 +148,34 @@ type EgressRule struct {
+// * ToCIDRSet
+// * ToFQDNs
+func (e *EgressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
```
Add a comment indicating why we are not appending `ToEndpoints` here.

jrajahalme replied:
Added.

ianvernon commented on pkg/policy/api/ingress.go:
```diff
@@ -108,12 +108,32 @@ type IngressRule struct {
+// * FromEntities
+// * FromCIDR
+// * FromCIDRSet
+func (i *IngressRule) SetAggregatedSelectors() {
```
Add a comment indicating why we are not appending `FromEndpoints` here.

jrajahalme replied:
Added.

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
 func (e *EgressRule) GetDestinationEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
+	if e.aggregatedSelectors == nil {
+		e.SetAggregatedSelectors()
```
There is no protection against parallel usage of `EgressRule` for this to happen safely.

jrajahalme replied:
It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

aanm replied:
this feels really odd... also, @jrajahalme comment not added :-)

ianvernon replied:
The comment was added in the implementation of `SetAggregatedSelectors`.

jrajahalme replied:
The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

aanm commented on pkg/policy/api/ingress.go:
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
 func (i *IngressRule) GetSourceEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
+	if i.aggregatedSelectors == nil {
+		i.SetAggregatedSelectors()
```
There is no protection against parallel usage of `IngressRule` for this to happen safely.

jrajahalme replied:
It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

aanm replied:
this feels really odd... also, @jrajahalme comment not added :-)

jrajahalme replied:
The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
+// ToEndpoints is not aggregated due to requirement folding in
+// GetDestinationEndpointSelectorsWithRequirements()
+func (e *EgressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
```
If To endpoints is not being used why the `len(e.ToEndpoints)` here?

jrajahalme replied:
I though I had fixed this, but apparently I've managed to lose those changes. Thanks for noticing!

aanm commented on pkg/policy/api/ingress.go:
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
+// FromEndpoints is not aggregated due to requirement folding in
+// GetSourceEndpointSelectorsWithRequirements()
+func (i *IngressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
```
If `FromEndpoints` is not being used why the `len(i.FromEndpoints) here

jrajahalme replied:
Fixed.

aanm commented on pkg/policy/api/egress.go:
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
 	// +optional
 	ToGroups []ToGroups `json:"toGroups,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
```
We are mixing the API with implementation details. This structure should kept somewhere else.

jrajahalme replied:
I fully agree. However, this was here before, and refactoring to an internal `IngressRule` and `EgressRule` is a bit too much for this cycle. Opened a new Issue #8353 to track this. Also added TODO comments here with reference to the GH issue.

aanm commented on pkg/policy/api/ingress.go:
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
 	// +optional
 	FromEntities EntitySlice `json:"fromEntities,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
```
Same here for the mix of API vs implementation details.

jrajahalme replied:
Same.

aanm commented on pkg/policy/api/ingress.go:
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
 	// +optional
 	FromEntities EntitySlice `json:"fromEntities,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
```
Same here for the mix of API vs implementation details.

jrajahalme replied:
Same.

jrajahalme committed:
policy: cache aggregated list of selectors in rule

Add a new, unexported field to the `IngressRule` and `EgressRule` types in the
API called `aggregatedSelectors`.

For `IngressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `FromEntities`
* `FromCIDR`
* `FromCIDRSet`

For `EgressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `ToEntities`
* `ToCIDR`
* `ToCIDRSet`
* `ToFQDNs`

Previously, upon each policy calculation, for each rule, this aggregate list of
rule fields as `EndpointSelector` was created in an ad-hoc manner, and was not
performant, as it allocated many different slices of `EndpointSelector` in calls
to `GetSourceEndpointSelectorsWithRequirements` and
`GetDestinationEndpointSelectorsWithRequirements`. This new field caches the
aggregated list only once, upon rule sanitization, or in calls to the two
aforementioned functions if the field has not been initialized yet.

The performance gains by making this change are significant. With 1000
egress rules containing a ToCIDR with 2 CIDRs, and 3000 identities
from before making this change:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     100	  12630576 ns/op	 4127697 B/op	  106148 allocs/op
```

And after:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     500	   6208546 ns/op	 2265445 B/op	   61134 allocs/op
```

This is a 50 % performance improvement in policy calculation for this simple
policy benchmark.

Signed-off-by: Ian Vernon <ian@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
